xdsl/parser.py

```diff
@@ -1,1335 +1,1924 @@
-        end_idx = self.idx
-        while self.file[end_idx] != '\n':
-            end_idx += 1
-        return self.file[start_idx:end_idx]
+        return cls(span.start, span.end, span.input)
 
+    @property
+    def string_contents(self):
+        # TODO: is this a hack-job?
+        return ast.literal_eval(self.text)
 
-@dataclass
-class ParserError(Exception):
-    """An error triggered during parsing."""
 
-    pos: Position | None
-    message: str
+@dataclass(frozen=True)
+class Input:
+    """
+    This is a very simple class that is used to keep track of the input.
+    """
+    content: str = field(repr=False)
+    name: str
 
-    def __str__(self):
-        if self.pos is None:
-            return f"Parsing error at end of file :{self.message}\n"
-        message = f"Parsing error at {self.pos}:\n"
-        message += self.pos.get_current_line() + '\n'
-        message += " " * (self.pos.column - 1) + "^\n"
-        message += self.message + '\n'
-        return message
+    @property
+    def len(self):
+        return len(self.content)
 
+    def __len__(self):
+        return self.len
 
-@dataclass
-class Parser:
+    def get_nth_line_bounds(self, n: int):
+        start = 0
+        for i in range(n):
+            next_start = self.content.find('\n', start)
+            if next_start == -1:
+                return None
+            start = next_start + 1
+        return start, self.content.find('\n', start)
+
+    def get_lines_containing(self,
+                             span: Span) -> tuple[list[str], int, int] | None:
+        # A pointer to the start of the first line
+        start = 0
+        line_no = 0
+        source = self.content
+        while True:
+            next_start = source.find('\n', start)
+            line_no += 1
+            # handle eof
+            if next_start == -1:
+                if span.start > len(source):
+                    return None
+                return [source[start:]], start, line_no
+            # as long as the next newline comes before the spans start we can continue
+            if next_start < span.start:
+                start = next_start + 1
+                continue
+            # if the whole span is on one line, we are good as well
+            if next_start >= span.end:
+                return [source[start:next_start]], start, line_no
+            while next_start < span.end:
+                next_start = source.find('\n', next_start + 1)
+            return source[start:next_start].split('\n'), start, line_no
 
-    class Source(Enum):
-        XDSL = 1
-        MLIR = 2
+    def at(self, i: int):
+        if i >= self.len:
+            raise EOFError()
+        return self.content[i]
 
-    ctx: MLContext
-    """xDSL context."""
 
-    str: str
-    """The current file/input to parse."""
+save_t = tuple[int, tuple[str, ...]]
+
 
-    source: Source = field(default=Source.XDSL, kw_only=True)
-    """The source language to parse."""
+@dataclass
+class Tokenizer:
```

Comment:
Could you document how this class is working?
Because currently, it is really hard to understand how the backtracking system works for instance.

Reply:
Will do!

Reply:
>Could you document how this class is working?
+1

Reply:
I added docs for this!